### PR TITLE
Jsonize hacksaw activity

### PR DESCRIPTION
**Pull request by Saicchi**
#### Summary
Infrastructure "Jsonize terrain and furniture hacksaw result"

#### Purpose of change
Hacksaw result in `activity_handlers::oxytorch_finish` is hardcoded, jsonizing it will allow easy addition of more terrain which it can be used and also new items which can use the hacksaw activity.

#### Describe the solution
Add a new field for terrain and furniture, `hacksaw` which is an object.
```cpp
"hacksaw": {
    "result": "terID/furnID", // (mandatory/optional) terrain/furniture it will become when done, defaults to f_null if furniture, otherwise it is mandatory
    "duration": "1 seconds", // ( optional ) time required for hacksawing, default is 1 second
    "message": "You finish cutting the metal.", // ( optional ) message that will be displayed when finished
    "byproducts": [ // ( optional ) list of items that will be spawned when finished
        {
            "item": "item_id",
            "count": 100 // exact amount
        },
        {
            "item": "item_id",
            "count": [ 10, 100 ] // random number in range ( inclusive )
        }
    ]
}
```

#### Describe alternatives you've considered
Keep hardcoding the results instead.

#### Testing
- [x] Hacksaw terrain
- [x] Hacksaw furniture
- [x] Spawns correct amount of byproducts
- [x] Multiple byproducts

#### Additional context
This is essentially a copy and paste of #50133 but modified for the hacksaw.

|                                                | result                      | duration   | byproducts                 |
| :---                                           | :---                        | :---       | :---                       |
| f_rack                                         | f_null                      | 2_minutes  | pipe[1,3], steel_chunk[1]  |
| t_chainfence                                   | t_dirt                      | 10_minutes | pipe[6], steel_chunk[20]   |
| t_chaingate_c                                  | t_dirt                      | 10_minutes | pipe[6], steel_chunk[20]   |
| t_chaingate_l                                  | t_dirt                      | 10_minutes | pipe[6], steel_chunk[20]   |
| t_chainfence_posts                             | t_dirt                      | 2_minutes  | pipe[6]                    |
| t_window_bars_alarm                            | t_window_alarm              | 10_minutes | rebar[1,8]                 |
| t_window_bars_curtains                         | t_window_domestic           | 10_minutes | rebar[1,8]                 |
| t_window_bars_domestic                         | t_window_domestic           | 10_minutes | rebar[1,8]                 |
| t_window_bars                                  | t_window_empty              | 10_minutes | rebar[1,8]                 |
| t_window_enhanced                              | t_window_reinforced         | 5_minutes  | rebar[1,4]                 |
| t_window_enhanced_noglass                      | t_window_reinforced_noglass | 5_minutes  | rebar[1,4]                 |
| t_reb_cage                                     | t_pit                       | 10_minutes | spike[19], scrap[8]        |
| t_retractable_gate_c                           | t_strconc_floor             | 10_minutes | chain[1,2], wire[8,22]     |
| t_retractable_gate_l                           | t_strconc_floor             | 10_minutes | chain[1,2], wire[8,22]     |
| t_bars                                         | t_floor                     | 15_minutes | pipe[3]                    |
| t_door_bar_c                                   | t_mdoor_frame               | 15_minutes | pipe[12]                   |
| t_door_bar_locked                              | t_mdoor_frame               | 15_minutes | pipe[12]                   |
| t_metal_grate_window                           | t_window_reinforced         | 10_minutes | pipe[1,12], sheet_metal[4] |
| t_metal_grate_window_with_curtain              | t_window_reinforced         | 10_minutes | pipe[1,12], sheet_metal[4] |
| t_metal_grate_window_with_curtain_open         | t_window_reinforced         | 10_minutes | pipe[1,12], sheet_metal[4] |
| t_metal_grate_window_noglass                   | t_window_reinforced_noglass | 10_minutes | pipe[1,12], sheet_metal[4] |
| t_metal_grate_window_with_curtain_noglass      | t_window_reinforced_noglass | 10_minutes | pipe[1,12], sheet_metal[4] |
| t_metal_grate_window_with_curtain_open_noglass | t_window_reinforced_noglass | 10_minutes | pipe[1,12], sheet_metal[4] |

When there is no `result` and the player finishes the hacksaw action, an error will show up on screen and the action won't finish.

Some terrains have special conditions, like `t_bars` which turns into `t_sewage` if around sewage tiles, otherwise it turns into `t_floor`. Best way to replicate it would be to have two identical `t_bars`, one just sewers.

### CODE REVIEW

**Comment by Saicchi** [comment]
Rebase to current experimental.

**Comment by BrettDong** [comment]
Need to remove unused `itype_id` in C++ code.

File: src/activity_actor.cpp [hunk 1]
```diff
@@ -1009,6 +1009,155 @@ std::unique_ptr<activity_actor> hotwire_car_activity_actor::deserialize( JsonIn
...
+
+void hacksaw_activity_actor::serialize( JsonOut &jsout ) const
+{
+    jsout.start_object();
+    jsout.member( "target", target );
+    jsout.member( "tool", tool );
+    jsout.end_object();
+}
+
+std::unique_ptr<activity_actor> hacksaw_activity_actor::deserialize( JsonIn &jsin )
```
Comment (kevingranade) [inline src/activity_actor.cpp:1152]:
```suggestion
std::unique_ptr<activity_actor> hacksaw_activity_actor::deserialize( JsonValue &jsin )
```

File: src/activity_actor_definitions.h [hunk 2]
```diff
@@ -281,6 +281,41 @@ class gunmod_remove_activity_actor : public activity_actor
...
+        void start( player_activity &act, Character &who ) override;
+        void do_turn( player_activity &/*act*/, Character &who ) override;
+        void finish( player_activity &act, Character &who ) override;
+
+        std::unique_ptr<activity_actor> clone() const override {
+            return std::make_unique<hacksaw_activity_actor>( *this );
+        }
+
+        void serialize( JsonOut &jsout ) const override;
+        static std::unique_ptr<activity_actor> deserialize( JsonIn &jsin );
```
Comment (kevingranade) [inline src/activity_actor_definitions.h:303]:
```suggestion
        static std::unique_ptr<activity_actor> deserialize( JsonValue &jsin );
```

File: src/activity_handlers.cpp [hunk 3]
```diff
@@ -193,16 +192,10 @@ static const itype_id itype_log( "log" );
 static const itype_id itype_mind_scan_robofac( "mind_scan_robofac" );
 static const itype_id itype_muscle( "muscle" );
 static const itype_id itype_nail( "nail" );
-static const itype_id itype_pipe( "pipe" );
-static const itype_id itype_rebar( "rebar" );
-static const itype_id itype_scrap( "scrap" );
-static const itype_id itype_sheet_metal( "sheet_metal" );
 static const itype_id itype_spike( "spike" );
```
Comment (kevingranade) [inline src/activity_handlers.cpp:195]:
```suggestion
```

File: tests/player_activities_test.cpp [hunk 4]
```diff
@@ -785,6 +794,272 @@ TEST_CASE( "boltcut", "[activity][boltcut]" )
...
+TEST_CASE( "hacksaw", "[activity][hacksaw]" )
+{
+    map &mp = get_map();
+    avatar &dummy = get_avatar();
+
+    auto setup_dummy = [&dummy]() -> item_location {
+        item it_hacksaw( itype_test_hacksaw );
+
+        dummy.wield( it_hacksaw );
+        REQUIRE( dummy.weapon.typeId() == itype_test_hacksaw );
```
Comment (kevingranade) [inline tests/player_activities_test.cpp:806]:
```suggestion
        REQUIRE( dummy.get_wielded_item().typeId() == itype_test_hacksaw );
```

File: tests/player_activities_test.cpp [hunk 5]
```diff
@@ -785,6 +794,272 @@ TEST_CASE( "boltcut", "[activity][boltcut]" )
...
+        REQUIRE( dummy.max_quality( qual_SAW_M ) == 10 );
+
+        return item_location{dummy, &dummy.weapon};
```
Comment (kevingranade) [inline tests/player_activities_test.cpp:809]:
```suggestion
        return item_location{dummy, &dummy.get_wielded_item()};
```

File: tests/player_activities_test.cpp [hunk 6]
```diff
@@ -785,6 +794,272 @@ TEST_CASE( "boltcut", "[activity][boltcut]" )
...
+            REQUIRE( mp.furn( tripoint_zero ) == furn_t_test_f_hacksaw3 );
+
+            item battery( itype_test_battery_disposable );
+            battery.ammo_set( battery.ammo_default() );
+
+            item it_hacksaw_elec( itype_test_hacksaw_elec );
+            it_hacksaw_elec.put_in( battery, item_pocket::pocket_type::MAGAZINE_WELL );
+
+            dummy.wield( it_hacksaw_elec );
+            REQUIRE( dummy.weapon.typeId() == itype_test_hacksaw_elec );
```
Comment (kevingranade) [inline tests/player_activities_test.cpp:933]:
```suggestion
            REQUIRE( dummy.get_wielded_item().typeId() == itype_test_hacksaw_elec );
```

File: tests/player_activities_test.cpp [hunk 7]
```diff
@@ -785,6 +794,272 @@ TEST_CASE( "boltcut", "[activity][boltcut]" )
...
+            REQUIRE( dummy.max_quality( qual_SAW_M ) == 10 );
+
+            item_location hacksaw_elec{dummy, &dummy.weapon};
```
Comment (kevingranade) [inline tests/player_activities_test.cpp:936]:
```suggestion
            item_location hacksaw_elec{ dummy, &dummy.get_wielded_item() };
```

**Pull request merged**